File: C++/210.cpp
__________________________________________________________________________________________________
sample 12 ms submission
static auto _=[]() {
    std::ios::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();
class Solution {
public:
    static const int MAXN = 5000;
    int head[MAXN],cnt = 0;
    int vis[MAXN],in[MAXN];
    void init() {
	    cnt = 0;
        memset(in,0,sizeof(in));
	    memset(head,-1,sizeof(head));
        memset(vis,0,sizeof(vis));
    }
    struct Edge {
	    int from;
	    int to;
	    int next;
    } edge[MAXN*2];
    void add(int from,int to) {
	    edge[cnt].from = from;
	    edge[cnt].to = to;
	    edge[cnt].next = head[from];
	    head[from] = cnt++;
    }
    
    vector<int> findOrder(int numCourses, vector<pair<int, int>>& prerequisites) {
        vector<int> ans;
        init();
        int n = prerequisites.size();
        for(int i = 0;i < n;++i)
        {
            add(prerequisites[i].second,prerequisites[i].first);
        }
        for(int i = 0;i < n;++i)
        {
            ++in[prerequisites[i].first];
        }
        queue<int> bfs;
        for(int i = 0;i < numCourses;++i)
        {
            if(in[i]==0)
                bfs.push(i);
        }
        while(!bfs.empty())
        {
            int node = bfs.front();
            bfs.pop();
            ans.push_back(node);
            for(int i = head[node];i!=-1;i=edge[i].next)
            {
                --in[edge[i].to];
                if(in[edge[i].to]<=0)
                    bfs.push(edge[i].to);
                if(edge[i].from!=node)
                    break;
            }
        }
        if(ans.size()!=numCourses)
            return vector<int>();
        return ans;
    }
};
__________________________________________________________________________________________________
sample 11008 kb submission
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<pair<int, int>>& prerequisites) {
        vector<int> ans;
        if(numCourses == 0)
            return ans;
        
        vector<int> pre(numCourses);
        for(auto p : prerequisites)
            pre[p.first]++;
        stack<int> stk;
        for(int i=0;i<numCourses;i++)
            if(pre[i] == 0)
                stk.push(i);
        while(!stk.empty()){
            int cur = stk.top();
            stk.pop();
            ans.push_back(cur);
            for(auto p : prerequisites){
                if(p.second == cur){
                    pre[p.first]--;
                    if(pre[p.first] == 0)
                        stk.push(p.first);
                }
            }
        }
        return ans.size() == numCourses ? ans : vector<int>();
    }
};
__________________________________________________________________________________________________

File: Python3/210.py
__________________________________________________________________________________________________
sample 100 ms submission
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj = [[] for _ in range(numCourses)]
        degree = [0] * numCourses
        for pre in prerequisites:
            adj[pre[1]].append(pre[0])
            degree[pre[0]] += 1
        queue = []
        for i in range(len(degree)):
            if degree[i] == 0:
                queue.append(i)
        res = []
        while queue:
            # get the no input directed node
            curr = queue.pop(0)
            # for the curr node adjacent nodes
            for node in adj[curr]:
                degree[node] -= 1
                if degree[node] == 0:
                    queue.append(node)
            res.append(curr)
            numCourses -= 1
        return res if numCourses == 0 else []
__________________________________________________________________________________________________
sample 14072 kb submission
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj = [[] for _ in range(numCourses)]
        indegree = {}
        for a, b in prerequisites:
            adj[b].append(a)
            indegree[a] = indegree.get(a, 0) + 1
        topological_sorted_order = []
        queue = collections.deque([])
        for node in range(numCourses):
            if node not in indegree:
                queue.append(node)
        while queue:
            node = queue.popleft()
            topological_sorted_order.append(node)
            for neighbor in adj[node]:
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    queue.append(neighbor)
        
        return topological_sorted_order if len(topological_sorted_order) == numCourses else []
__________________________________________________________________________________________________
